src/Features/CSharp/Portable/ConvertToRecord/ConvertToRecordHelpers.cs

```diff
@@ -297,43 +280,83 @@ public static (ImmutableArray<ArgumentSyntax> initializerArguments, ImmutableArr
                 {
                     // found a valid assignment to the parameter, add the expression it was assigned to
                     // to the initializer list, and add the eniter statement to be deleted
-                    initializerBuilder.Add(SyntaxFactory.Argument(assignmentValues[assignmentIndex].right));
-                    removalBuilder.Add(bodyOps[assignmentIndex].Syntax);
+                    initializerBuilder.Add(SyntaxFactory.Argument(assignmentValue.right));
+                    // don't need to remove argument syntaxes as the base initializer will already get removed
+                    if (assignmentValue.right.Parent is not ArgumentSyntax)
+                    {
+                        removalBuilder.Add(assignmentValue.right.GetAncestor<ExpressionStatementSyntax>()!);
+                    }
                 }
             }
 
             return (initializerBuilder.AsImmutable(), removalBuilder.AsImmutable());
         }
 
         private static ImmutableArray<(ISymbol left, T right)> GetAssignmentValuesForConstructor<T>(
-            ImmutableArray<IOperation> operations,
+            IConstructorBodyOperation constructorOperation,
             Func<IOperation, T?> captureAssignedSymbol)
-        => operations.SelectAsArray<IOperation, (ISymbol, T)>(operation =>
         {
-            if (operation is IExpressionStatementOperation
+            var body = GetBlockOfMethodBody(constructorOperation);
+            var _ = ArrayBuilder<(ISymbol left, T right)>.GetInstance(out var builder);
+
+            // We expect the constructor to have exactly one statement per property,
+            // where the statement is a simple assignment from the parameter's property to the property
+            if (body == null)
+            {
+                return builder.ToImmutable();
+            }
+
+            if (constructorOperation.Initializer is
+                IInvocationOperation { Arguments: ImmutableArray<IArgumentOperation> args })
+            {
+                foreach (var arg in args)
                 {
-                    Operation: ISimpleAssignmentOperation
+                    if (arg is { Parameter: IParameterSymbol param, Value: IOperation value } &&
+                        captureAssignedSymbol(value) is T captured)
                     {
-                        Target: IOperation assignee,
-                        Value: IOperation assignment
+                        var positionalParam = param.ContainingSymbol.ContainingType.GetMembers().FirstOrDefault(member
+                                => member.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax() ==
+                                    param.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax());
```

Comment:
Can we get a comment or two on this bit? Especially why we're comparing the syntax like this?